crates/ra_parser/src/lib.rs

```diff
@@ -31,12 +31,26 @@ pub struct ParseError(pub String);
 ///
 /// Hopefully this will allow us to treat text and token trees in the same way!
 pub trait TokenSource {
+    fn current(&self) -> Token;
+
+    /// Lookahead n token
+    fn lookahead_nth(&self, n: usize) -> Token;
```

Comment:
Originally it should be `lookahead_1` , `lookahead_2`, `lookahead_3` , but composite-able tokens and DOLLARS makes it impossible.